feat: custom parameters template substitution.

ivanvgh opened the pull request:
### Description:

This PR adds the capability to add custom parameter template values in order to be substituted by a processor.

We must have the following configuration in order to make this happen

- Add the following settings variable to map our template values with their corresponding processors. Also, these processors must return a string object.
`LTI_CUSTOM_PARAM_TEMPLATES = { 'templated_param_value': 'customer_package.module:func' }`

- In Studio we will have the custom parameters field value such as:
 `["static_param=static_value", "dynamic_custom_param=${templated_param_value}"]`

   **Note.** It must be wrapped by **${** and  **}** to indicate it is a templated_value to substitute.

### Use Case:
I want to be able to set dynamic custom parameters in the LTI consumer Xblock in Studio so that dynamic value should be resolved/substituted with a value resolved by the processor and added to the launch URL in order to be used by the provider as needed.

In order to make this happen, we must map the dynamic value template with the custom parameter processor in the LTI_CUSTOM_PARAM_TEMPLATES just like mentioned above, the custom parameter processor must be defined in the customer package.

edx-webhook commented:
Thanks for the pull request, @ivanvgh! I've created [OSPR-6283](https://openedx.atlassian.net/browse/OSPR-6283) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

natabene commented:
@ivanvgh Thank you for your contribution. Please let me know once this is ready for our review.

ivanvgh commented:
> @ivanvgh Thank you for your contribution. Please let me know once this is ready for our review.

Hi @natabene, this is ready for review, let me know if you have any questions, please.

ivanvgh commented on lti_consumer/utils.py:
```diff
@@ -113,3 +118,35 @@ def get_lti_nrps_context_membership_url(lti_config_id):
+
+            return template
+    except ValueError:
+        log.error(
```
@schenedx, you are right, I am applying your suggestion.

ivanvgh commented on lti_consumer/utils.py:
```diff
@@ -113,3 +118,35 @@ def get_lti_nrps_context_membership_url(lti_config_id):
+    :param xblock: LTI consumer xblock.
+    :param template: processor key.
+    """
+    template_value = template
```
@schenedx, you are right, I am applying your suggestion.

ivanvgh commented on lti_consumer/lti_xblock.py:
```diff
@@ -100,6 +100,7 @@
     'staff': 'Administrator',
     'instructor': 'Instructor',
 }
+CUSTOM_PARAMETER_TEMPLATE_TAG = '&'
```
@schenedx, I am using a pair of tags which are: **${** and **}** to wrap the **template_value** I checked in the LTI docs that they use **$** sign as the prefix for custom substitution, and in order to avoid collision with it we can use the pair of tags mentioned before, what do you think about it?

ivanvgh commented:
> I think I understand why this would enhance the custom templating on this repository. However, I'd like to know your use case better so I can understand how this is a benefit to the xblock-lti-consumer. Also, are you confident that the invocation of this functionality is safe when no `LTI_CUSTOM_PARAM_TEMPLATES` is defined? Could that be shown with at test case when `LTI_CUSTOM_PARAM_TEMPLATES` is not defined?

Hi @schenedx, thank you for reviewing this PR, I added a [test case](https://github.com/Pearson-Advance/xblock-lti-consumer/blob/ivanvgh/custom-params-template/lti_consumer/tests/unit/test_lti_xblock.py#L1623) that shows that it is safe to have this feature with the LTI_CUSTOM_PARAM_TEMPLATES not defined. Also, I've updated the PR description with a use case that will give a better explanation about this enhancement.

schenedx reviewed (approved):
👍 This looks good to me. Let me see what I can do to help you merge this PR.

schenedx commented:
@ivanvgh I need your help to make this PR releasable. Could you please update the [changelog at the bottom of the readme](https://github.com/openedx/xblock-lti-consumer/blob/master/README.rst). Then please update the [version of this repo](https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/__init__.py#L7) to something like 3.2.0. I think this your change prompts a minor version bump. Once that's done, I'll merge the PR and help you create a new release on the repo. I'll also help you upgrade the version of this library on [edx-platform](https://github.com/openedx/edx-platform/blob/master/requirements/edx/base.txt#L617). Deal?

ivanvgh commented:
> @ivanvgh I need your help to make this PR releasable. Could you please update the [changelog at the bottom of the readme](https://github.com/openedx/xblock-lti-consumer/blob/master/README.rst). Then please update the [version of this repo](https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/__init__.py#L7) to something like 3.2.0. I think this your change prompts a minor version bump. Once that's done, I'll merge the PR and help you create a new release on the repo. I'll also help you upgrade the version of this library on [edx-platform](https://github.com/openedx/edx-platform/blob/master/requirements/edx/base.txt#L617). Deal?

Sure, already done @schenedx, thank you so much.

edx-webhook commented:
@ivanvgh 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/xblock-lti-consumer&entry.867055334=https://github.com/openedx/xblock-lti-consumer/pull/215&entry.1484655318=https://github.com/ivanvgh&entry.752974735=2021-12-15+17:27&entry.1917517419=2022-01-19+14:48&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
